File: 20-valid-parentheses/valid-parentheses.ts
function isValid(s: string): boolean {
    const map = new Map<string, string>([
        ['(', ')'],
        ['{', '}'],
        ['[', ']']
    ]);

    const stack = new Array<string>();

    for (let char of s) {
        if (map.has(char)) stack.push(char);
        else {
            if (map.get(stack[stack.length - 1]) !== char) return false;
            else stack.pop();
        }
    }

    return stack.length === 0;
};

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> mp = new HashMap<>();
        mp.put('(', ')');
        mp.put('{', '}');
        mp.put('[', ']');
    
        Stack<Character> st = new Stack<>();

        for(int i = 0; i < s.length(); i ++) {
            Character curChar = s.charAt(i);
            if(mp.containsKey(curChar)) st.push(curChar);
            else{
                if(st.size() == 0) return false;
                Character topChar = st.pop();
                if(mp.get(topChar) != curChar) return false;
            }
        }
        return st.size() == 0;
    }
}